### PR TITLE
Rule to check every reference is used in body and footer

### PR DESCRIPTION
**Pull request by parhamsaremi**
Prohibit contributors from defining a reference and not putting it in the footer or putting a reference in the footer which is not defined in the body.

### CODE REVIEW

File: commitlintplugins.test.ts [hunk 1]
```diff
@@ -212,6 +212,34 @@ test('footer-notes-misplacement-3', () => {
...
+        + "\n\n"+ "Bla bla blah."
+        + "\n\n" + "[1] http://foo.bar/baz"
+    let footerReferenceExistence2 = runCommitLintOnMsg(commmitMsgwithWrongFooter);
+    expect(footerReferenceExistence2.status).not.toBe(0);
+})
+
+
+test('footer-references-existence3', () => {
+    let commmitMsgwithWrongFooter = "foo: this is only a title"
+        + "\n\n"+ "Bla bla blah[1], and after that [2], then [3]."
```
Comment (knocte) [inline commitlintplugins.test.ts:235]: @parhamsaremi let's add one more test, in which [4] is used, but there's no footer note for it.

**Comment by knocte** [comment]
@parhamsaremi why not squash the 2nd commit with the 1st?

File: commitlintplugins.test.ts (hunk 1 above)
Comment (parhamsaremi) [reply commitlintplugins.test.ts:235]: I didn't get this. don't we already have [2] which is used in body but there is no footer for it? :thinking:

**Comment by parhamsaremi** [comment]
> @parhamsaremi why not squash the 2nd commit with the 1st?

sure :+1:

File: commitlintplugins.test.ts (hunk 1 above)
Comment (knocte) [reply commitlintplugins.test.ts:235]: oh true, ok, then let's change this test, I prefer the number not being used being the last one, not the one in the middle

File: commitlint.config.ts [hunk 2]
```diff
@@ -397,6 +398,47 @@ module.exports = {
...
+                        let bodyStr = convertAnyToString(body, "body");
+                        let lines = bodyStr.split(/\r?\n/);
+                        let bodyReferences = new Set();
+                        let references = new Set();
+                        for (let line of lines) {
+                            let matches = line.match(/(?<=\[)([0-9]+)(?=\])/g);
+                            if (matches === null) {
+                                continue;
+                            }
+                            for (let m of matches){
```
Comment (knocte) [inline commitlint.config.ts:414]: let's rename m

Comment (parhamsaremi) [reply commitlint.config.ts]: done :+1:

**Comment by knocte** [comment]
You're missing an area in each commit message (it's easy to find which area you should use just by looking at the history of the commits in this repo). Don't forget next time please.

**Comment by parhamsaremi** [comment]
> You're missing an area in each commit message (it's easy to find which area you should use just by looking at the history of the commits in this repo). Don't forget next time please.

Oh sorry. Ok sure, I'll keep this in mind.

**Pull request merged**